covigator/references/lineage_annotation.py

```diff
@@ -1,27 +1,468 @@
+                    aa=first_aa,
+                    pos=protein_position
+                )
+                ref_aa = first_aa
+
+        return {"hgvs_p": hgvs,
+                "reference": ref_aa,
+                "alternate": "del",
+                "position": protein_position}
+
+    def get_hgvs_from_nuc_snp(self, variant) -> dict:
+        """
+        Translate pangolin nucleotide level SNPs description into hgvs protein annotation. Returns the variant
+        as hgvs string and the annotation of this variant
+
+        Args:
+            variant (dict): A dictionary with keys for genomic position, reference and alternate base
+
+        Returns:
+            annotation (dict): A dictionary with the annotation and HGVS notation of the SNP
+        """
+        position = variant.get("genomic_position")
+        ref_base = variant.get("reference")
+        alt_base = variant.get("alternate")
+        annotation = "synonymous"
+        gene_annot = self.find_gene(position)
+        # If mutation does not overlap any gene --> Intergenic
+        if gene_annot[0] is None:
+            annotation = "intergenic"
+            return {"hgvs_p": None,
+                    "position": position,
+                    "reference": ref_base,
+                    "alternate": alt_base,
+                    "annotation": annotation}
+
+        position_in_cds = position - (gene_annot[1] - 1)
+        protein_position = position_in_cds // 3 + 1 if position_in_cds % 3 != 0 else position_in_cds // 3
+        # Get wild type and mutated codon sequence
+        if position_in_cds % 3 == 0:
+            wt_aa = self.genome[(position - 2) - 1: position]
+            mt_aa = wt_aa[0:2] + alt_base
+        elif position_in_cds % 3 == 1:
+            wt_aa = self.genome[position - 1: position + 2]
+            mt_aa = alt_base + wt_aa[1::]
+        else:
+            wt_aa = self.genome[(position - 1) - 1: position + 1]
+            mt_aa = wt_aa[0] + alt_base + wt_aa[2]
+        # Translate and check determine type of mutation
+        wt_aa = str(Seq(wt_aa).translate())
+        mt_aa = str(Seq(mt_aa).translate())
+        if wt_aa != mt_aa:
+            annotation = "missense"
+        hgvs_p = "p.{ref}{aa_pos}{alt}".format(
+            ref=wt_aa,
+            aa_pos=protein_position,
+            alt=mt_aa
+        )
+        return {"hgvs_p": hgvs_p,
+                "position": protein_position,
+                "reference": wt_aa,
+                "alternate": mt_aa,
+                "annotation": annotation}
+
+    def _parse_mutation_sites(self, variant_string: str) -> List[Dict]:
+        """
+        Parse mutation site definition from pango constellation files. This code
+        is based on the parser found in the utility script of scorpio. Mutations
+        on nucleotide level are parsed and translated to protein level.
+        """
+        snp_pattern = re.compile('([ACTGUN]+)([0-9]+)([ACTGUN]+)')
```

Comment:
Do we want to support Ns? I don't know if there are any mutations described with N, but if there are they won't match any of our calls. I think we can kick this out.

For Us we could do a similar reasoning. But in this case, if they are described in pango lineages with U we should translate it to Ts in order to match with our mutations. I don't know if there are mutations described with U though.